### PR TITLE
optimize fgMorphTree for GenTreeArgList

### PR DESCRIPTION
**Pull request by sandreenko**
We can't usually have deep trees, because importer restricts trees depth:
https://github.com/dotnet/coreclr/blob/da40cb1ca8779566ca53ed6bd2fa67a926b29b38/src/jit/importer.cpp#L10173-L10178

However, it doesn't restrict depth of `GT_LIST` trees, so we can hit a stack overflow if there is a method with huge number of arguments.

This PR adds non-recursive morph for `GenTreeArgList`.

Fixes DevDiv_635180.

No asm diffs, no measurable instruction retired diff for System.Private.CoreLib.

### CODE REVIEW

**Comment by sandreenko** [comment]
To repro the issue we need:
1. call with many late arguments;
2. one of the arguments should be a CSE candidate.

Then inside 'PerformCSE' we call `fgMorphBlockStmt /* re-morph the statement */` and it calls `call->gtCallLateArgs = fgMorphTree(call->gtCallLateArgs)->AsArgList();` that causes stack overflow.

I am trying to create such test that would repro the issue in the checked mode, but have difficulties with creating many late arguments. Maybe somebody knows how to do that?

**Comment by sandreenko** [comment]
@dotnet-bot test Windows_NT x64 Checked jitstress2
@dotnet-bot test Windows_NT x86 Checked jitstress2

**Comment by sandreenko** [comment]
PTAL @dotnet/jit-contrib

File: src/jit/morph.cpp [hunk 1]
```diff
@@ -18888,3 +18891,64 @@ bool Compiler::fgCheckStmtAfterTailCall()
...
+//
+// Return Value:
+//    morphed argument list.
+//
+GenTreeArgList* Compiler::fgMorphArgList(GenTreeArgList* args, MorphAddrContext* mac)
+{
+    // Use non-recursive algorithm that morph all actual list values,
+    // memorize the last node for each effect flag and reset
+    // them during the second iteration.
+    constexpr int      numberOfTrackedFlags               = 5;
```
Comment (erozenfeld) [inline src/jit/morph.cpp]: morph --> morphs, memorize --> memorizes, reset --> resets

File: src/jit/morph.cpp [hunk 2]
```diff
@@ -18888,3 +18891,64 @@ bool Compiler::fgCheckStmtAfterTailCall()
...
+//    args - argument list tree to morph;
+//    mac  - morph address context, used to morph children.
+//
+// Return Value:
+//    morphed argument list.
+//
+GenTreeArgList* Compiler::fgMorphArgList(GenTreeArgList* args, MorphAddrContext* mac)
+{
+    // Use a non-recursive algorithm that morphs all actual list values,
+    // memorizes the last node for each effect flag and resets
```
Comment (CarolEidt) [inline src/jit/morph.cpp]: FWIW, "memoize" is a term often used in compilers.

**Comment by briansull** [comment]
> I am trying to create a test that would repro the issue in the checked mode, but have difficulties with creating a call with many late arguments. Maybe somebody knows how to do that?

Try using argumebt that have side-effects:  like:
 a[i]  // array bound check side-effect
o.f,  // null check o
t1=a,  // assignment to t1

**Pull request merged**